dynamiq/nodes/tools/python_code_executor.py

```diff
@@ -0,0 +1,459 @@
+        return None
+
+    @staticmethod
+    def _file_info_to_dict(file_info: FileInfo | None) -> dict[str, Any]:
+        if not file_info:
+            return {}
+        return {
+            "name": file_info.name,
+            "path": file_info.path,
+            "size": file_info.size,
+            "content_type": file_info.content_type,
+            "created_at": file_info.created_at.isoformat(),
+            "metadata": file_info.metadata,
+        }
+
+
+class PythonCodeExecutorInputSchema(BaseModel):
+    """Inputs for the PythonCodeExecutor tool."""
+
+    code: str = Field(..., description="Python code to execute. Must define a 'run' function.")
+    params: dict[str, Any] = Field(
+        default_factory=dict,
+        description="Optional parameters forwarded to the executed code.",
+    )
+    files: list[Any] | FileMappedInput | None = Field(
+        default=None,
+        description="Files available to the executed code.",
+        json_schema_extra={"map_from_storage": True},
+    )
+
+    model_config = ConfigDict(arbitrary_types_allowed=True, extra="allow")
+
+
+class PythonCodeExecutor(Node):
+    """Execute ad-hoc Python code inside RestrictedPython with file store helpers."""
+
+    group: Literal[NodeGroup.TOOLS] = NodeGroup.TOOLS
+    name: str = "PythonCodeExecutor"
+    description: str = """Runs dynamic Python code safely via RestrictedPython. Your snippet must define a `run(...)`
+function that acts as the entrypoint. The tool injects helper functions such as read_file(), write_file(),
+list_files(), and exposes the shared file store so you can read/write artifacts inside the workspace.
+Always return structured results from `run` — any `print()` output is captured separately in a `stdout` field."""
+    file_store: FileStore = Field(..., description="File storage backend shared with the agent.")
+    is_files_allowed: bool = True
+    input_schema: ClassVar[type[PythonCodeExecutorInputSchema]] = PythonCodeExecutorInputSchema
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    def execute(
+        self, input_data: PythonCodeExecutorInputSchema, config: RunnableConfig = None, **kwargs
+    ) -> dict[str, Any]:
+        code_preview = input_data.code.strip().replace("\n", "\\n")[:200]
+        logger.info(
+            f"Tool {self.name} - {self.id}: started with code length={len(input_data.code)} "
+            f"preview='{code_preview}{'...' if len(input_data.code) > 200 else ''}'"
+        )
+        config = ensure_config(config)
+        self.run_on_node_execute_run(config.callbacks, **kwargs)
+
+        if not self.file_store:
+            raise ToolExecutionException("PythonCodeExecutor requires a configured file_store.", recoverable=False)
+
+        file_manager = FileWorkspace(self.file_store)
+        helpers, helper_aliases = self._build_file_helpers(file_manager)
+
+        stdout_collector = PrintCollector()
+
+        def safe_print(*args, **print_kwargs):
```

Comment:
Do we need to define `safe_print` and `guarded_write` in each `execute` or we can define them on class/file level?